spec/lib/union_of_spec.rb

```diff
@@ -0,0 +1,1244 @@
+                  "licenses"
+                WHERE
+                  "licenses"."user_id" = '#{user.id}'
+              )
+              UNION
+              (
+                SELECT
+                  "licenses"."id"
+                FROM
+                  "licenses"
+                  INNER JOIN "license_users" ON "licenses"."id" = "license_users"."license_id"
+                WHERE
+                  "license_users"."user_id" = '#{user.id}'
+              )
+            ) "licenses"
+        SQL
+
+        expect(queries.second).to match_sql <<~SQL.squish
+          SELECT
+            "machine_components".*
+          FROM
+            "machine_components"
+            INNER JOIN "machines" ON "machine_components"."machine_id" = "machines"."id"
+            INNER JOIN "licenses" ON "machines"."license_id" = "licenses"."id"
+          WHERE
+            "licenses"."id" IN ('#{license.id}')
+          ORDER BY
+            "machine_components"."created_at" ASC
+        SQL
+      end
+    )
+  end
+
+  it 'should produce a deeper union join' do
+    expect(Product.joins(:users).to_sql).to match_sql <<~SQL.squish
+      SELECT
+        "products".*
+      FROM
+        "products"
+        INNER JOIN "policies" ON "policies"."product_id" = "products"."id"
+        INNER JOIN "licenses" ON "licenses"."policy_id" = "policies"."id"
+        INNER JOIN (
+          (
+            SELECT
+              "license_users"."user_id"    AS id,
+              "license_users"."license_id" AS union_id
+            FROM
+              "license_users"
+          )
+          UNION
+          (
+            SELECT
+              "licenses"."user_id" AS id,
+              "licenses"."id"      AS union_id
+            FROM
+              "licenses"
+          )
+        ) "users_union" ON "users_union"."union_id" = "licenses"."id"
+        INNER JOIN "users" ON "users"."id" = "users_union"."id"
+      ORDER BY
+        "products"."created_at" ASC
+    SQL
+  end
+
+  it 'should produce a deeper union query' do
+    product = create(:product, account:)
+
+    expect(product.users.to_sql).to match_sql <<~SQL.squish
+      SELECT
+        DISTINCT "users".*
+      FROM
+        "users"
+        INNER JOIN (
+          (
+            SELECT
+              "licenses"."id"           AS id,
+              "license_users"."user_id" AS union_id
+            FROM
+              "licenses"
+              INNER JOIN "license_users" ON "licenses"."id" = "license_users"."license_id"
+          )
+          UNION
+          (
+            SELECT
+              "licenses"."id"      AS id,
+              "licenses"."user_id" AS union_id
+            FROM
+              "licenses"
+          )
+        ) "licenses_union" ON "licenses_union"."union_id" = "users"."id"
+        INNER JOIN "licenses" ON "licenses"."id" = "licenses_union"."id"
+        INNER JOIN "policies" ON "licenses"."policy_id" = "policies"."id"
+      WHERE
+        "policies"."product_id" = '#{product.id}'
```

Comment:
This query is very slow when tested in production on accounts with >1M licenses and users.

We should consider denormalizing `product_id` onto `licenses`.

Reply:
Is there a way to push the joins up into the `UNION` derived table, so that we don't select all licenses for an account, only to discard ones that aren't for the correct policy? I think that's the root issue here.

Reply:
Using a `LATERAL` join works here in speeding up the query (from 800ms to 40ms), but it doesn't work on every through association (e.g. `Machine#users`). Need to figure out why it works here for `Product#users` but not on other associations and try to come up with a heuristic.

```sql
SELECT
  DISTINCT "users".*
FROM
  "users"
  INNER JOIN LATERAL (
    (
      SELECT
        "licenses"."id" AS id,
        "license_users"."user_id" AS union_id
      FROM
        "licenses"
        INNER JOIN "license_users" ON "licenses"."id" = "license_users"."license_id"
      WHERE
        "license_users"."user_id" = "users"."id"
        AND "licenses"."product_id" = '36b7e27e-b7ec-45c3-88a5-9cef0fb3acfa'
        AND "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
        AND "license_users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
    )
    UNION
    (
      SELECT
        "licenses"."id" AS id,
        "licenses"."user_id" AS union_id
      FROM
        "licenses"
      WHERE
        "licenses"."user_id" = "users"."id"
        AND "licenses"."product_id" = '36b7e27e-b7ec-45c3-88a5-9cef0fb3acfa'
        AND "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
    )
  ) "licenses_union" ON "licenses_union"."union_id" = "users"."id"
  INNER JOIN "licenses" ON "licenses"."id" = "licenses_union"."id"
WHERE
  "licenses"."product_id" = '36b7e27e-b7ec-45c3-88a5-9cef0fb3acfa'
  AND "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
ORDER BY
  "users"."created_at" DESC
LIMIT
  10;
```

Reply:
For the latter, this query performs better:

```sql
-- ~0.07 ms for 500k licenses, users, and machines
SELECT
  "users".*
FROM
  "users"
  INNER JOIN (
    (
      SELECT
        "licenses"."id" AS id,
        "license_users"."user_id" AS union_id
      FROM
        "licenses"
        INNER JOIN "license_users" ON "licenses"."id" = "license_users"."license_id"
      WHERE
        "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
        AND "license_users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
    )
    UNION
    (
      SELECT
        "licenses"."id" AS id,
        "licenses"."user_id" AS union_id
      FROM
        "licenses"
      WHERE
        "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
    )
  ) "licenses_union" ON "licenses_union"."union_id" = "users"."id"
  INNER JOIN "licenses" ON "licenses"."id" = "licenses_union"."id"
WHERE
  "licenses"."id" = '7ad79de7-2caa-45db-8bf3-73314f559f1e'
  AND "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
ORDER BY
  "users"."created_at" DESC
LIMIT
  101;
```

Reply:
We could also go super simple when we have a primary key:

```sql
(
  SELECT
    DISTINCT "users".*
  FROM
    "users"
    JOIN "license_users" ON "license_users"."user_id" = "users"."id"
  WHERE
    "license_users"."license_id" = '8c04deed-806f-4823-a7fc-6fcae940b678'
    AND "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
)
UNION
(
  SELECT
    DISTINCT "users".*
  FROM
    "users"
    JOIN "licenses" ON "licenses"."user_id" = "users"."id"
  WHERE
    "licenses"."id" = '8c04deed-806f-4823-a7fc-6fcae940b678'
    AND "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893';
)
```

Reply:
For joins, we could use the CTE, and for relations we could use the separate `UNION`s above.

Reply:
Can't do the simplified `UNION` because without the nested limit, the query performs badly for large tables. Pagination is also a challenge, since limit and offset will not be accurate.